File: tools/tf-types-ts/src/core/crypto.ts
/**
 * Crypto primitives — thin wrappers over audited pure-JS libraries.
 *
 * Supported:
 *   - ed25519 signing / verifying (via @noble/ed25519).
 *   - SHA-256 and BLAKE3 hashing (via @noble/hashes).
 *
 * Post-quantum ML-DSA is reserved in the SignatureEnvelope schema and
 * added in Phase 3+ behind a feature flag. No custom crypto is introduced
 * in this module.
 */

import * as ed from "@noble/ed25519";
import { sha256 } from "@noble/hashes/sha2";
import { blake3 } from "@noble/hashes/blake3";

// The async APIs (signAsync / verifyAsync / getPublicKeyAsync) use the Web
// Crypto SHA-512 implementation and do not need a sha512 callback. We only
// use the async APIs in this module.

export class CryptoError extends Error {}

export interface Ed25519KeyPair {
  readonly privateKey: Uint8Array;
  readonly publicKey: Uint8Array;
}

/** Generate a fresh ed25519 key pair from a provided 32-byte seed or Web Crypto randomness. */
export async function ed25519Generate(seed?: Uint8Array): Promise<Ed25519KeyPair> {
  const privateKey = seed ?? crypto.getRandomValues(new Uint8Array(32));
  if (privateKey.length !== 32) {
    throw new CryptoError(`ed25519 seed must be 32 bytes, got ${privateKey.length}`);
  }
  const publicKey = await ed.getPublicKeyAsync(privateKey);
  return { privateKey: new Uint8Array(privateKey), publicKey };
}

/** Derive the public key for an ed25519 private key. */
export async function ed25519PublicKey(privateKey: Uint8Array): Promise<Uint8Array> {
  if (privateKey.length !== 32) {
    throw new CryptoError(`ed25519 private key must be 32 bytes, got ${privateKey.length}`);
  }
  return ed.getPublicKeyAsync(privateKey);
}

/** Sign a message with ed25519. Deterministic per RFC 8032. */
export async function ed25519Sign(message: Uint8Array, privateKey: Uint8Array): Promise<Uint8Array> {
  if (privateKey.length !== 32) {
    throw new CryptoError(`ed25519 private key must be 32 bytes, got ${privateKey.length}`);
  }
  return ed.signAsync(message, privateKey);
}

/**
 * Verify an ed25519 signature. Resolves to `true` on success, `false` on any
 * failure mode. This mirrors the Rust implementation which returns
 * Result<(), CryptoError>.
 */
export async function ed25519Verify(
  publicKey: Uint8Array,
  message: Uint8Array,
  signature: Uint8Array,
): Promise<boolean> {
  if (publicKey.length !== 32) return false;
  if (signature.length !== 64) return false;
  try {
    return await ed.verifyAsync(signature, message, publicKey);
  } catch {
    return false;
  }
}

/** SHA-256 of the input, returned as `"sha256:<hex>"`. */
export function sha256HashRef(bytes: Uint8Array): string {
  return "sha256:" + toHex(sha256(bytes));
}

/** BLAKE3 of the input, returned as `"blake3:<hex>"`. */
export function blake3HashRef(bytes: Uint8Array): string {
  return "blake3:" + toHex(blake3(bytes));
}

/** Parse `"sha256:<hex>"` back into `{algorithm, bytes}`. */
export function parseHashRef(s: string): { algorithm: string; bytes: Uint8Array } {
  const colon = s.indexOf(":");
  if (colon < 0) throw new CryptoError(`malformed hashref: ${s}`);
  const algorithm = s.slice(0, colon);
  const hex = s.slice(colon + 1);
  if (hex.length % 2 !== 0) throw new CryptoError("odd-length hex");
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    const v = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    if (Number.isNaN(v)) throw new CryptoError("non-hex character");
    bytes[i] = v;
  }
  return { algorithm, bytes };
}

export function toHex(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

export function fromHex(s: string): Uint8Array {
  if (s.length % 2 !== 0) throw new CryptoError("odd-length hex");
  const out = new Uint8Array(s.length / 2);
  for (let i = 0; i < out.length; i++) {
    const v = parseInt(s.slice(i * 2, i * 2 + 2), 16);
    if (Number.isNaN(v)) throw new CryptoError("non-hex character");
    out[i] = v;
  }
  return out;
}

export function b64encode(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}

export function b64decode(s: string): Uint8Array {
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function utf8encode(s: string): Uint8Array {
  return new TextEncoder().encode(s);
}

export function utf8decode(b: Uint8Array): string {
  return new TextDecoder().decode(b);
}

// ---------- X25519 ----------

import { x25519 } from "@noble/curves/ed25519";

export interface X25519KeyPair {
  readonly privateKey: Uint8Array;
  readonly publicKey: Uint8Array;
}

export function x25519Generate(seed?: Uint8Array): X25519KeyPair {
  const privateKey = seed ?? crypto.getRandomValues(new Uint8Array(32));
  if (privateKey.length !== 32) {
    throw new CryptoError(`x25519 seed must be 32 bytes, got ${privateKey.length}`);
  }
  const publicKey = x25519.getPublicKey(privateKey);
  return { privateKey: new Uint8Array(privateKey), publicKey };
}

export function x25519DiffieHellman(privateKey: Uint8Array, peerPublic: Uint8Array): Uint8Array {
  if (privateKey.length !== 32) throw new CryptoError("x25519 private must be 32 bytes");
  if (peerPublic.length !== 32) throw new CryptoError("x25519 public must be 32 bytes");
  return x25519.getSharedSecret(privateKey, peerPublic);
}

// ---------- HKDF-SHA256 ----------

import { hkdf as nobleHkdf } from "@noble/hashes/hkdf";

export function hkdfSha256(inputKey: Uint8Array, salt: Uint8Array, info: Uint8Array, outputLen: number): Uint8Array {
  return nobleHkdf(sha256, inputKey, salt, info, outputLen);
}

// ---------- ChaCha20-Poly1305-IETF ----------

import { chacha20poly1305 } from "@noble/ciphers/chacha";

export class AeadError extends Error {}

export function chacha20poly1305Encrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  plaintext: Uint8Array,
): Uint8Array {
  if (key.length !== 32) throw new AeadError("aead key must be 32 bytes");
  if (nonce.length !== 12) throw new AeadError("aead nonce must be 12 bytes");
  return chacha20poly1305(key, nonce, aad).encrypt(plaintext);
}

export function chacha20poly1305Decrypt(
  key: Uint8Array,
  nonce: Uint8Array,
  aad: Uint8Array,
  ciphertext: Uint8Array,
): Uint8Array {
  if (key.length !== 32) throw new AeadError("aead key must be 32 bytes");
  if (nonce.length !== 12) throw new AeadError("aead nonce must be 12 bytes");
  try {
    return chacha20poly1305(key, nonce, aad).decrypt(ciphertext);
  } catch (e) {
    throw new AeadError(`aead authentication failed: ${(e as Error).message}`);
  }
}

File: crates/tf-types/src/crypto.rs
//! Crypto primitives — thin wrappers over reviewed crates.
//!
//! Supported:
//!   - ed25519 signing / verifying (via `ed25519-dalek`).
//!   - SHA-256 and BLAKE3 hashing.
//!
//! Post-quantum ML-DSA is a Phase 3+ addition and is reserved in the
//! `SignatureEnvelope` schema today. No custom crypto is introduced in
//! this module — everything is a thin adapter.

use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use ed25519_dalek::{Signature, Signer as _, SigningKey, Verifier as _, VerifyingKey};
use sha2::{Digest, Sha256};

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum CryptoError {
    #[error("invalid ed25519 private key: expected 32 raw bytes, got {0}")]
    PrivateKeyLength(usize),
    #[error("invalid ed25519 public key: expected 32 raw bytes, got {0}")]
    PublicKeyLength(usize),
    #[error("invalid ed25519 signature: expected 64 raw bytes, got {0}")]
    SignatureLength(usize),
    #[error("signature verification failed")]
    BadSignature,
    #[error("invalid public key encoding")]
    BadPublicKey,
    #[error("invalid base64: {0}")]
    BadBase64(String),
    #[error("unknown algorithm: {0}")]
    UnknownAlgorithm(String),
    #[error("{0}")]
    Generic(String),
}

/// ED25519 signing key.
pub struct Ed25519Signer {
    inner: SigningKey,
}

impl Ed25519Signer {
    pub fn from_bytes(seed: &[u8; 32]) -> Self {
        Ed25519Signer {
            inner: SigningKey::from_bytes(seed),
        }
    }

    pub fn generate<R: rand::RngCore + rand::CryptoRng>(rng: &mut R) -> Self {
        Ed25519Signer {
            inner: SigningKey::generate(rng),
        }
    }

    pub fn public_key_bytes(&self) -> [u8; 32] {
        self.inner.verifying_key().to_bytes()
    }

    pub fn sign(&self, msg: &[u8]) -> [u8; 64] {
        self.inner.sign(msg).to_bytes()
    }
}

/// Verify an ed25519 signature.
pub fn ed25519_verify(public_key: &[u8], msg: &[u8], signature: &[u8]) -> Result<(), CryptoError> {
    let pk_bytes: &[u8; 32] = public_key
        .try_into()
        .map_err(|_| CryptoError::PublicKeyLength(public_key.len()))?;
    let sig_bytes: &[u8; 64] = signature
        .try_into()
        .map_err(|_| CryptoError::SignatureLength(signature.len()))?;
    let vk = VerifyingKey::from_bytes(pk_bytes).map_err(|_| CryptoError::BadPublicKey)?;
    let sig = Signature::from_bytes(sig_bytes);
    vk.verify(msg, &sig).map_err(|_| CryptoError::BadSignature)
}

/// SHA-256 of the input, returned as `"sha256:<hex>"`.
pub fn sha256_hashref(bytes: &[u8]) -> String {
    let digest = Sha256::digest(bytes);
    format!("sha256:{}", hex(&digest))
}

/// BLAKE3 of the input, returned as `"blake3:<hex>"`.
pub fn blake3_hashref(bytes: &[u8]) -> String {
    let digest = blake3::hash(bytes);
    format!("blake3:{}", hex(digest.as_bytes()))
}

pub fn hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{:02x}", b));
    }
    s
}

/// Parse `sha256:<hex>` back into raw bytes.
pub fn parse_hashref(s: &str) -> Result<(String, Vec<u8>), CryptoError> {
    let (algo, hex_part) = s
        .split_once(':')
        .ok_or_else(|| CryptoError::UnknownAlgorithm(s.to_owned()))?;
    let mut out = Vec::with_capacity(hex_part.len() / 2);
    let bytes = hex_part.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if i + 1 >= bytes.len() {
            return Err(CryptoError::BadBase64("odd hex length".into()));
        }
        let hi = from_hex(bytes[i]).ok_or_else(|| CryptoError::BadBase64("non-hex char".into()))?;
        let lo = from_hex(bytes[i + 1]).ok_or_else(|| CryptoError::BadBase64("non-hex char".into()))?;
        out.push((hi << 4) | lo);
        i += 2;
    }
    Ok((algo.to_owned(), out))
}

fn from_hex(b: u8) -> Option<u8> {
    match b {
        b'0'..=b'9' => Some(b - b'0'),
        b'a'..=b'f' => Some(b - b'a' + 10),
        b'A'..=b'F' => Some(b - b'A' + 10),
        _ => None,
    }
}

/// Base64 encode / decode helpers for signature payloads.
pub fn b64encode(bytes: &[u8]) -> String {
    B64.encode(bytes)
}

pub fn b64decode(s: &str) -> Result<Vec<u8>, CryptoError> {
    B64.decode(s.as_bytes())
        .map_err(|e| CryptoError::BadBase64(e.to_string()))
}

// ---------- X25519 ----------

pub struct X25519KeyPair {
    pub private: [u8; 32],
    pub public: [u8; 32],
}

pub fn x25519_generate<R: rand::RngCore + rand::CryptoRng>(rng: &mut R) -> X25519KeyPair {
    let secret = x25519_dalek::StaticSecret::random_from_rng(rng);
    let public = x25519_dalek::PublicKey::from(&secret);
    X25519KeyPair {
        private: secret.to_bytes(),
        public: public.to_bytes(),
    }
}

pub fn x25519_from_bytes(seed: &[u8; 32]) -> X25519KeyPair {
    let secret = x25519_dalek::StaticSecret::from(*seed);
    let public = x25519_dalek::PublicKey::from(&secret);
    X25519KeyPair {
        private: secret.to_bytes(),
        public: public.to_bytes(),
    }
}

pub fn x25519_diffie_hellman(private: &[u8; 32], peer_public: &[u8; 32]) -> [u8; 32] {
    let secret = x25519_dalek::StaticSecret::from(*private);
    let peer = x25519_dalek::PublicKey::from(*peer_public);
    secret.diffie_hellman(&peer).to_bytes()
}

// ---------- HKDF-SHA256 ----------

pub fn hkdf_sha256(input_key: &[u8], salt: &[u8], info: &[u8], output_len: usize) -> Vec<u8> {
    let hk = hkdf::Hkdf::<Sha256>::new(Some(salt), input_key);
    let mut out = vec![0u8; output_len];
    hk.expand(info, &mut out).expect("output_len <= 255*32");
    out
}

// ---------- ChaCha20-Poly1305-IETF ----------

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum AeadError {
    #[error("aead key must be 32 bytes")]
    BadKey,
    #[error("aead nonce must be 12 bytes")]
    BadNonce,
    #[error("aead authentication failed")]
    AuthFailed,
}

pub fn chacha20poly1305_encrypt(
    key: &[u8; 32],
    nonce: &[u8; 12],
    aad: &[u8],
    plaintext: &[u8],
) -> Vec<u8> {
    use chacha20poly1305::aead::{Aead, KeyInit, Payload};
    use chacha20poly1305::ChaCha20Poly1305;
    let cipher = ChaCha20Poly1305::new_from_slice(key).expect("32-byte key");
    cipher
        .encrypt(
            chacha20poly1305::Nonce::from_slice(nonce),
            Payload {
                msg: plaintext,
                aad,
            },
        )
        .expect("encrypt")
}

pub fn chacha20poly1305_decrypt(
    key: &[u8; 32],
    nonce: &[u8; 12],
    aad: &[u8],
    ciphertext: &[u8],
) -> Result<Vec<u8>, AeadError> {
    use chacha20poly1305::aead::{Aead, KeyInit, Payload};
    use chacha20poly1305::ChaCha20Poly1305;
    let cipher = ChaCha20Poly1305::new_from_slice(key).map_err(|_| AeadError::BadKey)?;
    cipher
        .decrypt(
            chacha20poly1305::Nonce::from_slice(nonce),
            Payload {
                msg: ciphertext,
                aad,
            },
        )
        .map_err(|_| AeadError::AuthFailed)
}
